2017 post-review changes

mlange05 opened the pull request:
This merge provides a set of changes intended to address the review comments of the paper. It also adds J. Hueckelheim as a co-author and sanitises some references.

I believe I have captured most reviewer comments, although I think we should aim to side-step a general FEM vs. FD discussion. I have added a few "overview references" to demonstrate that FD is still widely used in various contexts, and I have tried to explain why FD is simply "one valuable tool in the PDE toolbox" in a new "Limitations and Future Work" section.

Commit:
MLange: Addressing various review comments and adding citations:

* Added three example citations for FD usage in different fields
  (seismic, electromagnetics and CFD).
* Clarify that we limit numerical discussion and chose examples
  for illustration purposes.
* Add citation to Peiro and Sherwing for numerical CFD background
* Note on example results that results are expected, but not numerically
  advanced.
* Mention coupled PDEs in Laplace intro
* Specify L1 norm and metion implicit methods

Commit:
MLange: Adding J. Hueckelheim as co-author

Commit:
MLange: Some small reference sanitations

Commit:
MLange: Add a "Limitations and Future Work" section

Commit:
MLange: Addressing a few more reviewer comments and fixing typos.

Commit:
MLange: More detail on the adjoint test.

Commit:
MLange: More small tweaks

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -74,14 +78,19 @@ with existing open-source software.
+widely used in industrial and academic research, although they merely
+represent one of many approaches to solving PDEs [Baba16]_, [Liu09]_,
+[Rai91]_. In this paper we therefore limit our discussion of numerical
+methods to a minium and instead focus on the ease which these
```
english feels odd here?

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -74,14 +78,19 @@ with existing open-source software.
+[Rai91]_. In this paper we therefore limit our discussion of numerical
+methods to a minium and instead focus on the ease which these
+operators can be created symbolically. We give a brief overview of the
+design concepts and Devito's key features and demonstrate the Devito
```
the devito -> its  ?

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -74,14 +78,19 @@ with existing open-source software.
+operators can be created symbolically. We give a brief overview of the
+design concepts and Devito's key features and demonstrate the Devito
+API using a set of classic examples from computational fluid dynamics
+(CFD). Based on this, we will then highlight the use of Devito in an
```
Based on this -> Then
highlight -> discuss

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -74,14 +78,19 @@ with existing open-source software.
+design concepts and Devito's key features and demonstrate the Devito
+API using a set of classic examples from computational fluid dynamics
+(CFD). Based on this, we will then highlight the use of Devito in an
+example of a complex seismic inversion algorithms to demonstrate its
```
"demonstrate" is used many times before as well as after this instance. Maybe use "show" ?

jhueckelheim replied:
example of complex seismic inversion algorithm**s**
example of **a** complex seismic inversion algorithm

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -300,19 +317,27 @@ for 100 timesteps are depicted in Figures :ref:`fig2dconv` and
-for this problem is
+symbolic code. However, the previous example only required a single
+variable to be updated, while more complex operators might need to
+execute multiple expressions simultaneously to solve coupled PDEs or
```
simultaneously, for example to solve ...

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -375,14 +400,14 @@ problem dimensions and cause the expression to be executed over the
-object.
+creating a :code:`sympy.Eq` object that assigns either fixed values or
+a prescribed function, such as the utility symbol :code:`bc_right` in
+our example, along the left and right boundary of our domain. To
```
our domain -> the domain

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -460,10 +488,17 @@ efficiently define rigorous forward modelling and adjoint operators
-equation is defined as
+[Louboutin17a]_.
+
+In the following example we will demonstrate the generation of forward
```
demonstrate -> show ?

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -825,6 +863,36 @@ such as flop reduction techniques - all while maintaining full
+efficiency, which is of particular importance for inversion methods
+that require fast and robust high-order PDE solvers.
+
+The high-level interfaces provided by Devito are intended to create
```
drop high-level ?

FabioLuporini commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -825,6 +863,36 @@ such as flop reduction techniques - all while maintaining full
+presented examples with greater accuracy or achieve more realistic
+results. Nevertheless, finite difference methods play an important
+role and are widely used in academic and industrial research due to
+the relative ease of implementation and high computational
```
implementation, verification/validation and ...?

FabioLuporini reviewed (commented):
I honestly don't have significant comments to make; this all sounds extremely reasonable to me. 

I just left some (very) minor comments about the use of the language.

jhueckelheim commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -74,14 +78,19 @@ with existing open-source software.
-achieved by the auto-generated and optimised code.
+utilised for a much broader set of computational problems. Matrix-free
+stencil operators based on explicit finite difference schemes are
+widely used in industrial and academic research, although they merely
```
Sounds as if industry is using this only for research? How about one of these:
industrial practice and academic research
industry and academic research
science and industry

jhueckelheim commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -162,9 +171,13 @@ Fluid Dynamics Examples
+in this section have been chosen for their relative simplicity in
+order to concisely demonstrate the capabilities and API features of
+Devito. For a more complete discussion on numerical methods for fluid
+flows please refer to [Peiro05]_.
```
We have chosen the examples in this section for their simplicity to concisely demonstrate...

jhueckelheim commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -278,9 +291,13 @@ tutorial by initialising the data associated with the symbolic function
+the results show good agreement with the original tutorial, they do
+exhibit significant numerical errors. This indicates that a high-order
+scheme, as well as a coupled formulation of the advection equation,
+might be preferred to achieve more realistic results.
```
...with the original tutorial, they do not represent an accurate solution to the linear convection equation. In particular, the low order spatial discretisation introduces numerical diffusion that changes the shape of the hat into a smooth bump. This is a well-known issue that could be addressed with more sophisticated solver schemes as discussed in [LeVeque, R. J., "Numerical Methods for Conservation Laws", Birkhauser-Verlag (1992)].

_The underlying issue here is that the hat is discontinuous (hence non-differentiable), violating a basic assumption in FD. [The referenced book](https://pdfs.semanticscholar.org/1470/c6f43c769572c4cfc94ffc9c5710484ff1e5.pdf) is discussing this in a very accessible way on pages 8-11. We could avoid the problem by using a smooth initial condition, but then it's not Lorena Barba's tutorial any more._

jhueckelheim commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -825,6 +863,36 @@ such as flop reduction techniques - all while maintaining full
+
+The examples used in this paper have been chosen for their relative
+simplicity in order to concisely demonstrate the current features of
+the Devito API. Different numerical methods may be used to solve the
```
Déjà vu. Almost the same sentence is in section "Fluid Dynamics Examples".

mloubout commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -402,18 +427,21 @@ object.
+against the name of the original symbol used to create the operator.
+
+The convergence criterion for this example was chosen to be
+:math:`\Vert p \Vert ^{1} < 10^{-4}`. The according initial condition
```
A bit vague.

The convergence criterion for this example is defined as the relative error between two iterations and set to ...

mloubout commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -460,10 +488,17 @@ efficiently define rigorous forward modelling and adjoint operators
+correctness using the so-called *adjoint test*. This test, also knwon
+as *dot product test*, verifies that the implementation of an adjoint
+operator indeed computes the conjugate transpose of the forward
+operator. Conceptually this amounts to verifying that the adjoint
```
I am not a big fan of saying "reverse" when talking about adjoints. I would just remove this sentence.

mlange05 replied:
Hmmm, I actually quite like the "reverse of information flow" concept (I got it from P. Farrell's theory section on the libadjoint website), especially given that this paper is for a non-expert audience w. r. t. inversion methods. Is this not what the dot product test verifies, that `A^T(A(q)) = q`? Anyway, happy to drop it, if it's misleading...

mloubout replied:
That would be the inverse test:
` A^(-1)(A(q)) = q` and is usually not teue for the adjoint.

Adjoint test only test that it is adjoint which is why I don't like saying reverse as it is misleading.

mlange05 replied:
Ok, cool, I'll take that line out. Just out of interest though, how else would you explain in non-expert terms what it means for an operator to be the adjoint of another?

mloubout replied:
It's a math definition, there is not much interpretation (or at least I don't know one).

mlange05 replied:
But that's exactly my point; "it's the Hermitian-transpose" is not particularly intuitive to non-experts. "Reversing the flow of information" at least gives some idea of what one might want to use an adjoint for. But anyway. nitpicking, I guess...

mloubout commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -460,10 +488,17 @@ efficiently define rigorous forward modelling and adjoint operators
+[Louboutin17a]_.
+
+In the following example we will demonstrate the generation of forward
+and adjoint operators for the acoustic wave equation and verify their
```
Split and ref http://library.seg.org/doi/abs/10.1190/1.3238367 for more details on seismic inversion.

Full PDF at https://jean-virieux.obs.ujf-grenoble.fr/IMG/pdf/GPY_2009_VIRIEUX.pdf?

mlange05 replied:
Not entirely sure how you want to split the sentence. Can you please be more specific? Happy to add the Virieux reference though.

mloubout replied:
Can't remember .... looks good now. Still would add the reference to virieux in the previous paragraph for seismic inversion.

mloubout reviewed (commented):
Mostly good to me, a few minor comments

Commit:
MLange: Small language tweaks according to co-author comments

mlange05 commented:
Ok, I updated most of the small language tweaks. Still need to change the initial condition on the advection example and add links to (updated) tutorials.

Commit:
MLange: Dropping vague adjoint description and adding Virieux reference

Commit:
MLange: Updating the advection example with a smooth initial condition

mlange05 commented:
Ok, apart from links to the relevant jupyter notebooks on the webpage (waiting PR opesci/devito#287 to land), I believe we are ready to respond to the reviewers. I propose the following rebuttals - comments and suggestions are welcome, as always:

Review response
=============

We would like to thanks all reviewers for their very clear and insightful comments. In response we have revised several sections of the paper to clarify the respective points. We hope that these changes have both made the paper clearer and addressed the reviewer's concerns. Responses to individual reviews are provided below. 

Reviewer 1 (@wd15)
===============

> The main criticism of the paper is that the disadvantages of the
approach and limitations of using Devito to pose and solve PDEs is not
discussed and needs to be addressed.

> A separate section discussing the current limitations of Devito and
plans for how the developers might envision addressing some of these
issues would be useful. Some consideration should be made to address
readers that are trying to understand the scope and limitations of
Devito and whether they should invest the time in learning and using
Devito for their own unique PDE problems.

We agree that some of the natural limitations of Devito were not discussed sufficiently in our original submission. We have therefore added a "Limitations and Future Work" section to the final discussion to clearly express the current state of development and potential future features that interested users might expect. 
However, with regards to posing and solving PDEs in general, we would like to point out that the primary motivation for Devito is to enable the efficient implementations of finite difference solvers, in a similar fashion to embedded DSLs for finite element methods like FEniCS or Firedrake. As such, we regard Devito as one of several tools to efficiently solve PDEs, leaving the choice of numerical scheme to the user, as it is highly problem-dependent. We have adjusted the introduction to better capture this sentiment, since we believe that a complete discussion and comparison of numeric approaches to solving PDEs is beyond the scope of this paper. 

> The first example, "linear convection", does not highlight that the
solution in Fig. 2 is not a good solution to a traveling square
wave. It is a good example for demonstrating how to use Devito, but
is not a realistic scheme for practical solutions of the advection
equation. In general a higher order scheme is required as well as
a limiting scheme or other specialized scheme to preserve the
travelling wave. This issue should be discussed. Furthermore, a
discussion of the limitations or difficulties raised in using
Devito for more complex schemes is required.

We agree that the initial condition from the original tutorial does lead to an overly diffusive solution that is not ideal if presented on its own. We have therefore updated the example with a smooth initial condition to avoid the numerical error due to discontinuities in the original "hat function". We also added a small discussion on the effective diffusion due to numerical errors and a reference to the LeVeque textbook for further reading.

> Again, the second example is not a good example of using Devito
either as it is a steady state diffusion problem, which is much
better resolved with an implicit scheme that Devito does not
support (or does it?). This issue should at least be discussed. The
topic of explicit versus implicit should be raised in the paper as
well. Some discussion of how Devito might move towards supporting
implicit schemes would be useful.

We again highlight that the primary objective of the presented examples is the illustration of the Devito API and accept that an implicit scheme might be more appropriate in a realistic scientific context. We have adjusted the discussion of the second example to reflect that and highlight a possible future integration with non-linear solver libraries to enable implicit schemes in the future.

> The paper should discuss how Devito developers might address the
issue of posing boundary conditions symbolically rather than having
to deal with indexing issues.

A note on future symbolic extensions for boundary conditions, has been added to the Laplace example, as well as the Future Work section of the paper. We believe that further details on the potential implementation of generic abstractions for FD boundary conditions would be premature at this point, since  there is no overarching concensus in the community on this topic. 

> All three examples only involve one equation. The paper should address
the capabilities and limitations of using Devito for coupled PDEs.

Coupled equations are actually supported with the current Devito API, due to operators allowing multiple SymPy expressions as input. We have highlighted this fact in the description of the Laplace example, which relies on the same mechanism to add boundary conditions. Moreover, a complete example of the coupled Burgers' equation is available as a notebook on the website. However, due to the space limitations of this paper and the additional complexity of the Burgers' example we have chosen to demonstrate this mechanism using the steady-state Laplace problem instead.

> The paper would be better if there are clear links to full code
examples (in the text) for all three of the examples in the paper.
Also, Docker containers for all or each example would allow users to
run the examples in the paper in future when dependencies become
outdated.

The examples presented in this paper are provided as jupyter notebooks on the website. We have provided explicit links to each, as well as including them in the packaged version of the code attached to the updated Zenodo reference.

jhueckelheim commented:
We have adjusted to the introduction -> We have adjusted the introduction

mlange05 commented:
Reviewer 2 (@nicholasmalaya)
=========

> This is my greatest principle criticism of the paper, which does not provide an motivation for the use of FD versus FEM. I do not believe it is appropriate to assume that the SciPy audience will all appreciate trade-offs between numerical discretization schemes. Thus, I strongly recommend at least a paragraph providing motivation for the need of a finite difference library be provided in the introduction. A small table highlighting the differences (or pro/con) between frameworks would be highly interesting.

As pointed out above, we believe that a comparison of numerical schemes is highly problem-dependent and much too complex to be covered rigorously in this paper. Instead we regard Devito as an additional tool in the scientific Python ecosystem that is specific to a particular widely used numerical method. We have adjusted the introduction accordingly and added an explicit statement for this sentiment to the new Future Work section.

> Figure 4 states 'after convergence' but I do not see anywhere in the text where the convergence criterion is defined. Perhaps it is while l1norm > 1.e-4? Regardless, this should be explicitly noted in the text to assist users with reproducibility.

This section has been rewritten to explicitly explain and state the convergence criterion. 

> The axes, plot labels, and title of figure 6 and 7 are too small. The size is not readable.

The size of the figure has been increased.

> The vertical dotted line in plot 6 and 7 are labeled 'SO'. In general it is desirable for a plot to be largely self-documented, in that the figure caption should define what SO is, for instance. Otherwise the reader must flip between pages to discover what this means.

We have provided additional information on setup, including an explanation of the "SO" abbreviation in the caption.

> Finally, can the authors comment on distributed computing environments (likely in the conclusions)? Would it be possible to expand DEVITO to support this? MPI is supported in FEniCS, and while DEVITO supports a shared memory paradigm (openMP), a large swath of HPC users are interested in multi-node parallelism.

This is a very good point that was missed in the original submission. One of the key drivers behind the ongoing YASK integration, beyond the "stencil folding", is that YASK comes with built-in MPI support and hybrid MPI-OpenMP support. This is now mentioned in the Future Work section.

mlange05 commented:
Reviewer 3 (@cyrush)
=================

> I do think the other reviewers have valid points about FEM vs FD. I viewed this as an illustrative example for how to use Devito, not a definitive example for how to best solve the CFD PDEs, but I agree that should be explicitly stated as to not lead folks astray.

The use of the CFD examples as an illustration of the API was indeed the main intention of the paper, and we have tried to make point clearer through changes in the introduction and the new Limitations and Future Work section. We have also tried to underline that the choice of numerical method is a highly problem-dependent issue, a full exploration of which would go beyond the scope of this paper. 

> I also recommend adding a reference or some text to introduce the Seismic Inversion application and some context for the Adjoint Test would help.

We have tried to make the meaning of the adjoint test more explicit and added a reference to Virieux and Operto, 2009 for further background on its use in seismic inversion methods.

We have also tried to fix all further typos and text/label issues.

FabioLuporini commented:
> We have provide explicit l

We have provided

> much to complex to be covered

too complex to be covered

FabioLuporini commented:
Looks good to me

jhueckelheim commented:
much to complex to be covered rigorously -> much too complex to be covered rigorously
"This is actually a very good point" -> "This is a very good point" (otherwise it sounds to me as if you are contrasting the fact that this is a good point with all the other points).

navjotk commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -402,18 +434,21 @@ object.
+
+The convergence criterion for this example is defined as the relative
+error between two iterations and set to :math:`\Vert p \Vert ^{1} <
+10^{-4}`. The according initial condition and the resulting
```
according -> corresponding (or just remove the word)

navjotk commented on papers/michael_lange/michael_lange.rst:
```diff
@@ -402,18 +434,21 @@ object.
+:ref:`fig2dlaplacefinal` respectively, agree with the original
+tutorial implementation. It should again be noted that the chosen
+numerical scheme might not be optimal to solve steady-state problems
+of this type, since often implicit methods are preferrable.
```
since often implicit methods are preferrable -> since implicit methods are often preferred (or fix typo in preferable)

navjotk reviewed (approved):
Only minor suggestions here. 

I will try to follow this up with an edit to slightly change/expand the explanation of the adjoint test.

Commit:
MLange: Add explicit references to webpage tutorials and update Zenodo
